chainer/function_node.py

```diff
@@ -747,55 +746,57 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
     #    are ignored and their creators are not searched.
-    candidate_funcs = [v.creator_node for v in outputs
-                       if v.creator_node is not None]
-    visited_funcs = set()
-    forward_graph = collections.defaultdict(list)
-    while candidate_funcs:
-        func = candidate_funcs.pop()
-        if func in visited_funcs:
+    candidate_vars = [y.node for y in outputs]
+    output_nodes = list(candidate_vars)  # TODO(kataoka): remove this
+    visited_vars = set()
+    forward_var_to_func = collections.defaultdict(list)
+    forward_func_to_var = collections.defaultdict(list)
+    while candidate_vars:
+        y = candidate_vars.pop()
+        if y in visited_vars:
             continue
-        visited_funcs.add(func)
-        for x in func.inputs:
-            # Raise error here if x is created by Function.backward.
-            # In such case, we don't know exact inputs of the creator.
-            x._check_old_style_gradient()
+        visited_vars.add(y)
 
-            if not x.requires_grad:
-                continue
-            forward_graph[x].append(func)
-            creator = x.creator_node
-            if creator is not None and creator not in visited_funcs:
-                candidate_funcs.append(creator)
+        # Raise error here if y is created by Function.backward.
+        # In such case, we don't know exact inputs of the creator.
+        y._check_old_style_gradient()
+
+        func = y.creator_node
+        if func is None:
+            continue
+        if func not in forward_func_to_var:  # func is not visited
+            for x in func.inputs:
+                if x.requires_grad:
+                    candidate_vars.append(x)
+                    forward_var_to_func[x].append(func)
+        # always append to forward graph
+        forward_func_to_var[func].append(y)
 
     # 2. Forward enumeration: all the nodes in the subgraph reachable from the
     #    input nodes are enumerated. The extracted (sub-)subgraph is the union
     #    of all paths that backpropagation will visit.
     candidate_vars = [x.node for x in inputs]
+    visited_vars = set()
     visited_funcs = set()
-    grad_required = set()
     while candidate_vars:
         x = candidate_vars.pop()
-        grad_required.add(x)
-        for func in forward_graph[x]:
+        if x in visited_vars:
+            continue
+        visited_vars.add(x)
+        assert x in forward_var_to_func or x in output_nodes
+        for func in forward_var_to_func[x]:
             if func in visited_funcs:
                 continue
             visited_funcs.add(func)
-            for y_ref in func.outputs:
-                y = y_ref()
-                if y is not None and y in forward_graph:
-                    candidate_vars.append(y)
+            assert func in forward_func_to_var
+            for y in forward_func_to_var[func]:
+                candidate_vars.append(y)
```

Comment:
How about checking `y not in visited_vars` before `candidate_vars.append(y)` instead of checking after dequeuing?